lib/domain/usecase/create_room_interactor.dart

```diff
@@ -0,0 +1,49 @@
+import 'package:dartz/dartz.dart';
+import 'package:fluffychat/app_state/failure.dart';
+import 'package:fluffychat/app_state/success.dart';
+import 'package:fluffychat/di/global/get_it_initializer.dart';
+import 'package:fluffychat/domain/app_state/room/create_room_failed.dart';
+import 'package:fluffychat/domain/app_state/room/create_room_success.dart';
+import 'package:fluffychat/domain/app_state/room/set_room_avatar_failed.dart';
+import 'package:fluffychat/domain/app_state/room/set_room_avatar_success.dart';
+import 'package:fluffychat/domain/model/room/new_room_request.dart';
+import 'package:fluffychat/domain/usecase/set_room_avatar_interactor.dart';
+import 'package:matrix/matrix.dart';
+
+class CreateRoomInteractor {
+  final setRoomAvatarInteractor = getIt.get<SetRoomAvatarInteractor>();
+
+  Stream<Either<Failure, Success>> execute(Client matrixClient, NewRoomRequest newRoomInformations) async* {
+    try {
+      final roomId = await matrixClient.createGroupChat(
+        groupName: newRoomInformations.groupName,
+        invite: newRoomInformations.invite,
+        enableEncryption: newRoomInformations.enableEncryption,
+        preset: newRoomInformations.createRoomPreset,
+      );
+
+      final withAvatarUpload = newRoomInformations.avatar != null;
+
+      if (roomId.isNotEmpty) {
+        yield Right(CreateRoomSuccess(event: CreateRoomSuccessEventType.createRoomSuccess, roomId: roomId, isDone: !withAvatarUpload));
+
+        if (withAvatarUpload) {
+          final Stream setRoomAvatarStream = setRoomAvatarInteractor.execute(matrixClient, roomId, newRoomInformations.avatar);
+
+          await for (final Either<SetRoomAvatarFailed, SetRoomAvatarSuccess> result in setRoomAvatarStream) {
+            if (result.isRight()) {
+              yield Right(CreateRoomSuccess(event: CreateRoomSuccessEventType.setAvatarSuccess, roomId: roomId, isDone: true));
+            } else if (result.isLeft()) {
+              throw result.fold(
+                (left) => left.exception,
+                (right) => right
+              );

```

Comment:
hope you can find better solution to make it clearer.

Reply:
I'm working on it, trying to make it simpler but I'm getting out of ideas